File: src/quick_sort/partition_and_quick_sort.java
package quick_sort;

class Solution {
  public void swap(int[] arr, int i, int j) {
    int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
  }

  public int partition(int[] arr, int l, int r) {
    if (l > r)
      return -1;
    if (l == r)
      return l;
    int lessEqual = l - 1, index = l;
    while (index < r) {
      if (arr[index] <= arr[r])
        swap(arr, index, ++lessEqual);

      index++;
    }
    swap(arr, ++lessEqual, r);
    return lessEqual;
  }

  public int[] netherlandsFlag(int[] arr, int L, int R) {
    if (L > R) { // L...R L>R
      return new int[] { -1, -1 };
    }
    if (L == R) {
      return new int[] { L, R };
    }
    int less = L - 1; // < 区 右边界
    int more = R; // > 区 左边界
    int index = L;
    while (index < more) { // 当前位置，不能和 >区的左边界撞上
      if (arr[index] == arr[R]) {
        index++;
      } else if (arr[index] < arr[R]) {
        swap(arr, index++, ++less);
      } else { // >
        swap(arr, index, --more);
      }
    }
    swap(arr, more, R); // <[R] =[R] >[R]
    return new int[] { less + 1, more };
  }

  public void QuickSort1(int[] arr) {
    if (arr == null || arr.length < 2)
      return;
    process1(arr, 0, arr.length - 1);
  }

  private void process1(int[] arr, int l, int r) {
    if (l >= r)
      return;
    int mid = partition(arr, l, r);
    process1(arr, l, mid - 1);
    process1(arr, mid + 1, r);
  }

  public void QuickSort2(int[] arr) {
    if (arr == null || arr.length < 2)
      return;
    process2(arr, 0, arr.length - 1);
  }

  // arr[L...R]排有序，快排 划分区间
  private void process2(int[] arr, int l, int r) {
    if (l >= r)
      return;
    int[] equalArea = netherlandsFlag(arr, l, r);
    process2(arr, l, equalArea[0] - 1);
    process2(arr, equalArea[1] + 1, r);
  }

  public void QuickSort3(int[] arr) {
    if (arr == null || arr.length < 2)
      return;
    process3(arr, 0, arr.length - 1);
  }

  private void process3(int[] arr, int l, int r) {
    if (l >= r)
      return;
    swap(arr, l + (int) (Math.random() * (r - l + 1)), r);
    int[] equalArea = netherlandsFlag(arr, l, r);
    process3(arr, l, equalArea[0] - 1);
    process3(arr, equalArea[1] + 1, r);
  }

  // =============== TEST ===============
  public static int[] generateRandomArray(int maxSize, int maxValue) {
    int[] arr = new int[(int) ((maxSize + 1) * Math.random())];
    for (int i = 0; i < arr.length; i++) {
      arr[i] = (int) ((maxValue + 1) * Math.random()) - (int) (maxValue * Math.random());
    }
    return arr;
  }

  public static int[] copyArray(int[] arr) {
    if (arr == null) {
      return null;
    }
    int[] res = new int[arr.length];
    for (int i = 0; i < arr.length; i++) {
      res[i] = arr[i];
    }
    return res;
  }

  public static boolean isEqual(int[] arr1, int[] arr2) {
    if ((arr1 == null && arr2 != null) || (arr1 != null && arr2 == null)) {
      return false;
    }
    if (arr1 == null && arr2 == null) {
      return true;
    }
    if (arr1.length != arr2.length) {
      return false;
    }
    for (int i = 0; i < arr1.length; i++) {
      if (arr1[i] != arr2[i]) {
        return false;
      }
    }
    return true;
  }

  public static void printArray(int[] arr) {
    if (arr == null) {
      return;
    }
    for (int i = 0; i < arr.length; i++) {
      System.out.print(arr[i] + " ");
    }
    System.out.println();
  }

  public static void main(String[] args) {
    Solution solution = new Solution();
    int testTime = 500000, maxSize = 100, maxValue = 100;
    boolean succeed = true;
    for (int i = 0; i < testTime; i++) {
      int[] arr1 = generateRandomArray(maxSize, maxValue);
      int[] arr2 = copyArray(arr1);
      int[] arr3 = copyArray(arr1);
      solution.QuickSort1(arr1);
      solution.QuickSort2(arr2);
      solution.QuickSort3(arr3);
      if (!isEqual(arr1, arr2) || !isEqual(arr2, arr3)) {
        succeed = false;
        break;
      }
    }
    System.out.println(succeed ? "Nice!" : "Oops!");

  }
}

File: src/quick_sort/partition_and_quick_sort.cpp
#include <iostream>
using namespace std;
class Solution {
 private:
  void swap(vector<int>& arr, int i, int j) {
    int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
  }

  int partition(vector<int>& arr, int l, int r) {
    if (l > r) return -1;
    if (l == r) return l;
    int lessEqual = l - 1, index = l;
    while (index < r) {
      if (arr[index] <= arr[r]) swap(arr, index, ++lessEqual);
      index++;
    }
    swap(arr, ++lessEqual, r);
    return lessEqual;
  }

  vector<int> netherlandsFlag(vector<int>& arr, int l, int r) {
    if (l > r) return {-1, -1};
    if (l == r) return {l, r};

    int less = l - 1, more = r, index = l;
    while (index < more) {
      if (arr[index] == arr[r])
        index++;
      else if (arr[index] < arr[r])
        swap(arr, index++, ++less);
      else
        swap(arr, index, --more);
    }
    swap(arr, more, r);
    return {less + 1, more};
  }

  void process1(vector<int>& arr, int l, int r) {
    if (l >= r) return;
    int mid = partition(arr, l, r);
    process1(arr, l, mid - 1);
    process1(arr, mid + 1, r);
  }

  void process2(vector<int>& arr, int l, int r) {
    if (l >= r) return;
    vector<int> equal_area = netherlandsFlag(arr, l, r);
    process2(arr, l, equal_area[0] - 1);
    process2(arr, equal_area[1] + 1, r);
  }
  void process3(vector<int>& arr, int l, int r) {
    if (l >= r) return;
    swap(arr, l + (int)(rand() % (r - l + 1)), r);
    vector<int> equal_area = netherlandsFlag(arr, l, r);
    process3(arr, l, equal_area[0] - 1);
    process3(arr, equal_area[1] + 1, r);
  }

 public:
  void quick_sort(vector<int>& arr) {
    if (arr.empty() || arr.size() < 2) return;
    process1(arr, 0, arr.size() - 1);
  }

  void quick_sort2(vector<int>& arr) {
    if (arr.empty() || arr.size() < 2) return;
    process2(arr, 0, arr.size() - 1);
  }
  void quick_sort3(vector<int>& arr) {
    if (arr.empty() || arr.size() < 2) return;
    process3(arr, 0, arr.size() - 1);
  }

  // ========== TEST ==========
  vector<int> generate_random_array(int maxSize, int maxValue) {
    vector<int> arr((int)rand() % (maxSize + 1), 0);
    for (int i = 0; i < arr.size(); i++) {
      arr[i] = (int)((rand() % (maxValue + 1)) - (rand() % maxValue));
    }
    return arr;
  }

  vector<int> copy_array(vector<int>& arr) {
    vector<int> res(arr.size(), 0);
    for (int i = 0; i < arr.size(); i++) res[i] = arr[i];

    return res;
  }

  bool is_equal(vector<int>& arr1, vector<int>& arr2) {
    if (arr1.size() != arr2.size()) return false;

    for (int i = 0; i < arr1.size(); i++) {
      if (arr1[i] != arr2[i]) return false;
    }
    return true;
  }

  void print_array(vector<int>& arr) {
    for (int i = 0; i < arr.size(); i++) {
      cout << arr[i] << " ";
    }
    cout << endl;
  }
};

int main() {
  Solution solution;
  int testTime = 500000, maxSize = 100, maxValue = 100;
  bool succeed = true;

  for (int i = 0; i < testTime; i++) {
    vector<int> arr1 = solution.generate_random_array(maxSize, maxValue);
    vector<int> arr2 = solution.copy_array(arr1);
    vector<int> arr3 = solution.copy_array(arr1);
    solution.quick_sort(arr1);
    solution.quick_sort2(arr2);
    solution.quick_sort3(arr3);
    if (!solution.is_equal(arr1, arr2) || !solution.is_equal(arr2, arr3)) {
      solution.print_array(arr1);
      solution.print_array(arr2);
      solution.print_array(arr3);
      succeed = false;
      break;
    }
  }
  cout << (succeed ? "Nice!" : "Oops!") << endl;

  return 0;
}